### PR TITLE
add track selector

### PR DESCRIPTION
**Pull request by sofiatomassini**
@njacazio @fbellini @neelimaagrawal

### CODE REVIEW

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]
Thanks for bringing this functionality forward!!!

Apart from my in place suggested changes there is an overall structure which you have to follow.

At least for the time being, this is going to become the Femto3D framework if I'm not too wrong. Not quite sure if Femto3D is enough representative of the analyses that you will carry but for illustration purposes I will use this name. If you decide something more representative just use it instead of Femto3D.

You will need to place your code in a new directory which you will create as
```
O2Physics/PWGCF/Femto3D
```
then you have to modify the `O2Physics/PWGCF/CMakeLists.txt` to include your new sub-directory in the build tree.

Your sub-directory has to have the same structure as the own `PWGCF`. You can check the structure and replicate it by looking at the `O2Physics/PWGCF/TwoParticleCorrelations` or at the `O2Physics/PWGCF/MultiparticleCorrelations` sub-directories. 

Pay special attention to the `CMakeLists.txt` files that you have to incorporate in each of the sub-directories that you will create. You can copy them from any of the two suggested sub-directories and modify them according to your needs/filenames.

The way to place your files in your new sub-directory structure is the same that you chose for them at the `PWGCF` level but now at the `PWGCF/Femto3D` level.

File: PWGCF/DataModel/singletrackselector.h [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+template <typename binningType>
+float unPack(const typename binningType::binned_t& b)
+{
+  return static_cast<float>(binningType::bin_width * b);
+
+  // return static_cast<float>((binningType::binned_max - binningType::binned_min) * 0.5 + binningType::bin_width * b);
+}
+
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/DataModel/singletrackselector.h]: Indeed this is true but this is not completely the same, as we changed the strategy to pack and unpack. We also changed the binning and this was the rationale

File: PWGCF/DataModel/singletrackselector.h [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+ public:
+  typedef int8_t binned_t;
+  static constexpr int nbins = (1 << 8 * sizeof(binned_t)) - 2;
+  static constexpr binned_t overflowBin = nbins >> 1;
+  static constexpr binned_t underflowBin = -(nbins >> 1);
+  static constexpr float binned_max = 10.0;
+  static constexpr float binned_min = -10.0;
+  static constexpr float bin_width = (binned_max - binned_min) / nbins;
+};
+} // namespace nsigma
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/DataModel/singletrackselector.h]: thanks for spotting this, indeed we changed the binning in order to have greater granularity. The function is different from the one reported in PIDResponse.h We changed the name to better reflect the functionality of the function

File: PWGCF/DataModel/singletrackselector.h [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                           [](storedcrossedrows::binning::binned_t binned) -> float { return singletrackselector::unPackInt<storedcrossedrows::binning>(binned); });
+DECLARE_SOA_DYNAMIC_COLUMN(TOFNSigmaPr, tofNSigmaPr,
+                           [](nsigma::binning::binned_t nsigma_binned) -> float { return singletrackselector::unPack<nsigma::binning>(nsigma_binned); });
+DECLARE_SOA_DYNAMIC_COLUMN(TPCNSigmaPr, tpcNSigmaPr,
+                           [](nsigma::binning::binned_t nsigma_binned) -> float { return singletrackselector::unPack<nsigma::binning>(nsigma_binned); });
+DECLARE_SOA_DYNAMIC_COLUMN(TOFNSigmaDe, tofNSigmaDe,
+                           [](nsigma::binning::binned_t nsigma_binned) -> float { return singletrackselector::unPack<nsigma::binning>(nsigma_binned); });
+DECLARE_SOA_DYNAMIC_COLUMN(TPCNSigmaDe, tpcNSigmaDe,
+                           [](nsigma::binning::binned_t nsigma_binned) -> float { return singletrackselector::unPack<nsigma::binning>(nsigma_binned); });
+
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/DataModel/singletrackselector.h]: Thanks for spotting this, you have a good point, however, the strategy here is to store px, py and pz and derived p and pT from them as dynamic columns. This is different with what used centrally (where px py and pz are dynamic). For the PID the binning is different and we decided to keep greater granularity

File: PWGCF/DataModel/singletrackselector.h [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+{
+  if (valueToBin <= binningType::binned_min) {
+    return binningType::underflowBin;
+  } else if (valueToBin >= binningType::binned_max) {
+    return binningType::overflowBin;
+  } else if (valueToBin >= 0) {
+    return static_cast<typename binningType::binned_t>((valueToBin / binningType::bin_width) + 0.5f);
+  } else {
+    return static_cast<typename binningType::binned_t>((valueToBin / binningType::bin_width) - 0.5f);
+  }*/
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/DataModel/singletrackselector.h]: Thanks, we removed the commented part

File: PWGCF/TableProducer/singleTrackSelector.cxx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                 //  track.tpcNClsCrossedRows(),
+                 singletrackselector::packInTableInt<singletrackselector::storedcrossedrows::binning>(track.tpcNClsCrossedRows()),
+                 singletrackselector::packInTable<singletrackselector::nsigma::binning>(track.tofNSigmaPr()),
+                 singletrackselector::packInTable<singletrackselector::nsigma::binning>(track.tpcNSigmaPr()),
+                 singletrackselector::packInTable<singletrackselector::nsigma::binning>(track.tofNSigmaDe()),
+                 singletrackselector::packInTable<singletrackselector::nsigma::binning>(track.tpcNSigmaDe()));
+
+        if ((singletrackselector::unPackInt<singletrackselector::storedcrossedrows::binning>(singletrackselector::packInTableInt<singletrackselector::storedcrossedrows::binning>(track.tpcNClsCrossedRows())) != track.tpcNClsCrossedRows()) && (singletrackselector::unPackInt<singletrackselector::storedcrossedrows::binning>(singletrackselector::packInTableInt<singletrackselector::storedcrossedrows::binning>(track.tpcNClsCrossedRows())) != (track.tpcNClsCrossedRows() - 1))) {
+          LOG(info) << "crossedRows: " << track.tpcNClsCrossedRows()
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/TableProducer/singleTrackSelector.cxx]: We removed the log

File: PWGCF/TableProducer/singleTrackSelector.cxx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                    << ", Unpacked crossedRows: " << singletrackselector::unPackInt<singletrackselector::storedcrossedrows::binning>(singletrackselector::packInTableInt<singletrackselector::storedcrossedrows::binning>(track.tpcNClsCrossedRows()));
+          // << "deve essere uguale a" << singletrackselector::unPack<singletrackselector::nsigma::binning>;
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/TableProducer/singleTrackSelector.cxx]: Done

File: PWGCF/TableProducer/singleTrackSelector.cxx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        }
+      }
+    }
+
+    // LOG(info) << "UnPacked crossedRows: " << singletrackselector::unPack(o2::aod::singletrackselector::storedcrossedrows::binning::binned_t(o2::aod::singletrackselector::storedcrossedrows::binning(tpcNClsCrossedRows)) )
+    // LOG(info) << "Unpacked crossedRows: " << unpackedValue;
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/TableProducer/singleTrackSelector.cxx]: Done

File: PWGCF/Tasks/histogramtest.cxx [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+      registry.add("TPCandTOF_eta_stdcut", "TPCandTOF_eta_stdcut", kTH2F, {{100, 0., 20.0}, {200, -1., 1.}});
+      registry.add("hcrossedrows_stdcut", ";track crossed rows;entries", kTH1F, {{160, -0.5, 159.5}});
+      registry.add("charge_pt_stdcut", "charge_pt_stdcut", kTH2F, {{100, 0., 10.}, {10, -1.5, 1.5}});
+      registry.add("charge_stdcut", "charge_stdcut", kTH1F, {{100, -1.5, 1.5}});
+      registry.add("NsigmaTPC_d_stdcut", "NsigmaTPC_d_stdcut", kTH2F, {{100, 0., 5.0}, {100, -5., 5.0}});
+      registry.add("NsigmaTOF_d_stdcut", "NsigmaTOF_d_stdcut", kTH2F, {{100, 0., 5.0}, {100, -5., 5.0}});
+      registry.add("TOFSignalpr_stdcut", ";TOFSignalpr_stdcut", kTH2F, {{1000, 0., 5.0}, {200, 0., 200.}});
+      registry.add("TOFSignalK_stdcut", "TOFSignalK_stdcut", kTH2F, {{1000, 0., 5.0}, {200, 0., 200.}});
+      registry.add("TOFSignald_stdcut", "TOFSignald_stdcut", kTH2F, {{1000, 0., 5.0}, {200, 0., 200.}});
+    }
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/Tasks/histogramtest.cxx]: Done

File: PWGCF/Tasks/histogramtest.cxx [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+            if (track.hasTOF()) {
+              registry.fill(HIST("TPCandTOF_stdcut"), track.pt());
+              registry.fill(HIST("TPCandTOF_eta_stdcut"), track.pt(), track.eta());
+            }
+          }
+        }
+      }
+    }
+    PROCESS_SWITCH(EtaPhiHistograms, processStandard, "process non filtered track", false);
+  */
```
[the remark this answers is not in the document]
Comment (sofiatomassini) [reply PWGCF/Tasks/histogramtest.cxx]: Done

**Review by sofiatomassini — COMMENTED** [review]
Thanks @victor-gonzalez we took into consideration your useful comments and we simplified a lot the code that we propose to have merged. Please don't hesitate in case something can still be improved!

**Review by victor-gonzalez — APPROVED** [review]
Approved for the time being
Future iterations will require the move on the source code hierarchy

**Review by victor-gonzalez — APPROVED** [review]

**Comment by victor-gonzalez** [comment]
You shouldn't incorporate anything to a PR which has not been previously fully compiled and tested locally

**Comment by sofiatomassini** [comment]
> You shouldn't incorporate anything to a PR which has not been previously fully compiled and tested locally

We tested and compiled the code locally. The only thing that is changed is the name of the file singletrackselector.cxx  

The error in the build/O2Physics/o2 is due to the compiler, and I couldn't test it on my Mac.

**Review by victor-gonzalez — APPROVED** [review]

**Comment by victor-gonzalez** [comment]
The change in the source filename should have been caught by a local O2Physics production 
Regarding the include error, never include something which starts with '_' unless you saw it in an already existing source file in O2

**Comment by sofiatomassini** [comment]
> The change in the source filename should have been caught by a local O2Physics production Regarding the include error, never include something which starts with '_' unless you saw it in an already existing source file in O2

Thank you for your comment! This is my first pull request and I will pay more attention next time.

**Pull request merged**